File: app/static/js/screens/cash_position.js
/* §6.4's equation, term by term, and the decision that carries a person's name.
 *
 * ## Why every term is shown rather than a single figure
 *
 * `cash_position.py` returns all of them deliberately, and its docstring says why: "a manager
 * told 'you are ₹500 short' and nothing else cannot check the claim, and this is the number
 * that decides whether a debt lands on somebody's name." So this screen renders the equation
 * as a worked sum, in the order §6.4 states it, with each term labelled -- not a total with a
 * disclosure triangle.
 *
 * ## The two figures are never added together
 *
 *   accountable_cash   what the meters and the other channels say he should be holding
 *   declared_cash      what he says he counted into the locker
 *
 * §5.2 and §14: summing these double-counts the day. They are computed independently by the
 * server and only ever *subtracted*, and this screen keeps them visually separate for the
 * same reason.
 *
 * ## gap: null is not zero
 *
 * `gap: null` means nobody has declared, which is a different fact from a gap of ₹0.00. §6.8's
 * "zero as an answer, never zero as an omission" again -- and the booking control is disabled
 * with that reason rather than offered and then refused with 409 NO_CASH_DECLARED.
 *
 * ## A shortfall is booked by a human, never by this screen
 *
 * §14 forbids booking automatically, and §5.2 gives the argument with more force than
 * anywhere else in the document, because here the debt is explicit and carries a name: "a
 * ₹500 gap is more often a mistyped reading, a forgotten UPI figure or an unrecorded udhaar
 * slip than it is theft, and the software must not be the thing that decides."
 *
 * So the computed gap is *shown*, the booking form is pre-filled with it, and the amount
 * remains editable -- because §5.2 stores `computed_gap` and `amount` separately and expects
 * them to differ sometimes. A divergence logs a warning server-side and writes anyway; it
 * never refuses, because refusing a human's judgement sends the correction outside the system
 * where nothing can see it.
 */

import { el, pill, render } from "../dom.js";
import { api, explain, Submission } from "../api.js";
import { format, gapLabel, isZero } from "../money.js";
import { businessDate } from "../time.js";
import { field, Form } from "../ui/field.js";
import { openSheet } from "../ui/sheet.js";
import { notify } from "../ui/toast.js";
import { errorCard } from "./today.js";

export async function renderCashPosition(container, { session, navigate, shiftId }) {
  const { shell } = session;
  shell.setTab("cash");
  shell.setTitle("Cash position");
  shell.setActions();

  render(container, el("div", { className: "t-caption", text: "Loading…" }));

  let position;
  try {
    position = await api.get(`/shifts/${shiftId}/cash-position`);
  } catch (error) {
    render(
      container,
      errorCard(error, () => renderCashPosition(container, { session, navigate, shiftId })),
    );
    return;
  }

  shell.setTitle("Cash position", businessDate(position.business_date));

  const context = { session, container, navigate, shiftId, position };
  const gap = gapLabel(position.gap);
  const declared = position.declared_cash;

  render(
    container,
    el("div", { className: "stack" }, [
      // The comparison, kept as two figures side by side rather than one net number.
      el("div", { className: "card stack" }, [
        el("div", { className: "t-micro", text: "The comparison" }),
        el("div", { className: "list" }, [
          term("Accountable — what the meters imply", position.accountable_cash),
          declared === null
            ? el("div", { className: "list-row" }, [
                el("div", { className: "list-row-main t-body", text: "Declared — what he counted" }),
                el("div", { className: "list-row-value t-absent", text: "not declared" }),
              ])
            : term("Declared — what he counted", declared),
        ]),
        el("div", { className: "row-between", style: { marginTop: "0.75rem" } }, [
          el("span", { className: "t-micro", text: "Gap" }),
          el("span", { className: `t-title t-numeric ${gap.className}`, text: gap.text }),
        ]),
        el("p", {
          className: "t-caption",
          text:
            declared === null
              ? "Nobody has declared cash for this shift, so there is nothing to compare against. That is not the same as a gap of zero."
              : "Positive means short, negative means a surplus. Nothing here is written — a gap becomes a debt only when a manager books it.",
        }),
      ]),

      position.incomplete
        ? el("div", { className: "card" }, [
            pill("incomplete", "review"),
            el("p", {
              className: "t-caption",
              text: "Some nozzles have no closing reading yet, so the accountable figure is partial.",
            }),
          ])
        : null,

      // §6.4 written out. The order matches the equation in the spec exactly, so somebody
      // holding the document can follow along line by line.
      el("div", { className: "card stack" }, [
        el("div", { className: "t-micro", text: "How the accountable figure is built" }),
        el("div", { className: "list" }, [
          term("Metered fuel sales", position.metered_fuel_sales),
          term("Non-fuel sales", position.non_fuel_sales),
          divider("less what did not arrive as cash"),
          term("Card", position.card_total, true),
          term("UPI", position.upi_total, true),
          term("Wallet", position.wallet_total, true),
          term("Credit sales (udhaar)", position.credit_sales_total, true),
          divider("plus cash that arrived from elsewhere"),
          term("Credit repayments in cash", position.cash_credit_repayments),
          term("Shortfall settlements in cash", position.cash_shortfall_settlements),
          divider("less cash that left"),
          term("Cash expenses", position.cash_expenses, true),
        ]),
        el("p", {
          className: "t-caption",
          text: "Only cash-mode expenses and repayments appear here. A bank-paid bill is on the record but never leaves the drawer.",
        }),
      ]),

      el("div", { className: "card stack" }, [
        el("div", { className: "t-micro", text: "Booked against this shift" }),
        el("div", { className: "list" }, [
          term("Shortfalls booked", position.shortfalls_booked),
        ]),
        el("p", {
          className: "t-caption",
          text: "Subtracted from the day's expected closing, so the same money is not counted both as a debt and as cash in the locker.",
        }),
        bookingControl(position, context),
      ]),
    ]),
  );
}

function term(label, value, negative = false) {
  return el("div", { className: "list-row" }, [
    el("div", { className: "list-row-main t-body", text: label }),
    el("div", {
      className: "list-row-value t-body t-numeric",
      text: `${negative ? "− " : ""}${format(value)}`,
    }),
  ]);
}

function divider(text) {
  return el("div", { className: "list-row" }, [
    el("div", { className: "list-row-main t-micro", text }),
  ]);
}

function bookingControl(position, context) {
  // 409 NO_CASH_DECLARED is what the server returns when there is no declaration to compare
  // against. Disabling with the reason is better than offering a button that always fails.
  if (position.gap === null) {
    return el("p", {
      className: "t-caption",
      text: "Booking a shortfall needs a declared cash figure first — there is no gap to book against a blank.",
    });
  }

  if (isZero(position.gap)) {
    return el("p", {
      className: "t-caption",
      text: "The drawer balances. Nothing to book.",
    });
  }

  // A surplus is not a shortfall, and V1 has no record type for one. §14's open questions
  // list "is a surplus ever booked?" as still unanswered, so this says so rather than
  // offering a control that would put a negative amount somewhere it does not belong.
  if (position.gap.trim().startsWith("-")) {
    return el("p", {
      className: "t-caption",
      text: "This shift shows a surplus rather than a shortfall. V1 has no record type for a surplus — it stays visible here and in the day's variance.",
    });
  }

  return el("button", {
    className: "btn btn-primary btn-block",
    text: "Book a shortfall",
    attrs: { type: "button" },
    on: { click: () => bookingSheet(position, context) },
  });
}

function bookingSheet(position, context) {
  const amount = field({
    name: "amount",
    label: "Amount to book",
    type: "number",
    step: "0.01",
    min: "0.01",
    // Pre-filled with the computed gap, and deliberately still editable. §5.2 stores both
    // `computed_gap` and `amount` because a manager may know part of the gap is a slip he
    // has already corrected.
    value: position.gap,
    required: true,
    inputMode: "decimal",
    hint: `The system computed ${format(position.gap)}. You may book a different figure — both are stored.`,
  });

  const reason = field({
    name: "reason",
    label: "Why is this being booked?",
    required: true,
    hint: "Mandatory. This becomes a debt in a named person's ledger, so the reason is part of the record. 3–500 characters.",
  });

  const form = new Form({ amount, reason });

  const submit = el("button", {
    className: "btn btn-primary btn-block",
    text: "Book shortfall",
    attrs: { type: "button" },
  });

  const submission = new Submission("POST", `/shifts/${context.shiftId}/shortfalls`);

  submit.addEventListener("click", async () => {
    form.clearErrors();
    const values = form.values();

    if (!values.reason || values.reason.trim().length < 3) {
      form.showErrors([{ loc: ["body", "reason"], msg: "A reason is required." }]);
      return;
    }

    submit.disabled = true;
    try {
      // No salesman_id. §14: it is read from shifts.attendant_id -- the one name §5.2 says
      // carries the drawer -- and a client-supplied value would let a typo put a debt on the
      // wrong person with no second source of truth to catch it.
      await submission.run({ amount: values.amount, reason: values.reason.trim() });
      sheet.close();
      notify.success("Shortfall booked.");
      renderCashPosition(context.container, context);
    } catch (error) {
      submit.disabled = false;
      const unmatched = error.isValidation ? form.showErrors(error.detail) : [];
      if (!error.isValidation || unmatched.length) {
        notify.error(explain(error), { requestId: error.requestId });
      }
    }
  });

  const sheet = openSheet({
    title: "Book a shortfall",
    body: el("div", { className: "stack" }, [
      el("div", { className: "card stack" }, [
        el("p", {
          className: "t-caption",
          text: "This records a debt against the salesman who carried this shift's drawer. It is repaid in cash.",
        }),
        el("p", {
          className: "t-caption",
          text: "A gap is more often a mistyped reading, a forgotten UPI figure or an unrecorded udhaar slip than it is theft. Check those first.",
        }),
      ]),
      ...form.nodes(),
    ]),
    footer: el("div", { style: { padding: "0 1rem 1rem" } }, [submit]),
  });
}

File: app/api/v1/cash_position.py
"""A shift's cash position -- what the meters say against what the salesman says
(CLAUDE.md §5.2, §6.4, §8, §13.14).

Phase 10. One route, and it **writes nothing**.

## The two figures, and why they are never merged

§5.2 is explicit that a `cash` collection row is *not* a term in §6.4's equation. Cash is
derived as the residual; the declared row is the **independent observation the derived figure
is checked against**:

* **accountable_cash** -- what the meters and the other payment channels say the salesman
  should have handed over.
* **declared_cash** -- what he says he counted into the locker.

The gap between them is the shortfall this outlet books as udhaar against his own name
(§14). §14 also forbids summing the two, and nothing here does: they are computed
independently and only ever subtracted.

## Why it writes nothing

§4.7's argument, which carries more weight here than anywhere else in the document because
the output has a person's name on it: *"an assumed opening converts theft into a debt owed by
someone who did nothing wrong."* A ₹500 gap is more often a mistyped reading, a forgotten UPI
figure or an unrecorded udhaar slip than it is theft. So this endpoint computes and shows;
a **manager** books, with a reason (§13.14, Step 7).

## Role floor

Manager (§8). Attendants may read their own shift's data-entry sheets, but this is a report
about them -- and the person a shortfall would be booked against is the last one who should
be able to run the calculation privately before anybody else sees it. `collections.py`'s
permission tests already make that argument for the declared figure.
"""

from __future__ import annotations

from datetime import date
from decimal import Decimal
from uuid import UUID

from fastapi import APIRouter, Depends
from pydantic import BaseModel
from sqlalchemy.orm import Session

from app.api.deps import ShiftAccess, require_shift_access
from app.core.roles import Role
from app.db.session import get_db
from app.services import cash as cash_service

router = APIRouter(tags=["cash"])


class CashPositionResponse(BaseModel):
    """Every term of §6.4's per-shift half, plus the two figures that get compared.

    Returned in full rather than as a single number on purpose. §5.2's argument for storing
    `expected_closing`'s components applies to showing them too: a manager told "you are
    ₹500 short" and nothing else cannot check the claim, and this is the number that decides
    whether a debt lands on somebody's name.
    """

    shift_id: UUID
    business_date: date
    salesman_id: UUID

    metered_fuel_sales: Decimal
    non_fuel_sales: Decimal
    card_total: Decimal
    upi_total: Decimal
    wallet_total: Decimal
    credit_sales_total: Decimal
    cash_credit_repayments: Decimal
    card_upi_credit_repayments: Decimal
    cash_shortfall_settlements: Decimal
    cash_expenses: Decimal

    accountable_cash: Decimal
    declared_cash: Decimal | None
    gap: Decimal | None
    shortfalls_booked: Decimal
    incomplete: bool

    gap_basis: str = (
        "gap = accountable_cash - declared_cash. Positive means short, negative means a "
        "surplus, and null means nobody has declared yet -- which is NOT zero. Nothing "
        "here is written: a gap becomes a debt only when a manager books it (CLAUDE.md "
        "§13.14)."
    )
    margin_basis: str = (
        "Priced with rate_at only. No profit figure is computed here -- CLAUDE.md §6.4 "
        "needs what the fuel was worth, not what it earned, and a missing dealer margin "
        "must not make a day unreconcilable (§6.3)."
    )


@router.get("/shifts/{shift_id}/cash-position", response_model=CashPositionResponse)
def read_cash_position(
    access: ShiftAccess = Depends(require_shift_access(Role.manager)),
    db: Session = Depends(get_db),
) -> CashPositionResponse:
    """§6.4's per-shift figures. Manager floor; nothing is written.

    Deliberately **not** `writable=True`. The gap is only knowable once the shift is closed
    and its readings are final, so refusing a closed shift would make the endpoint unreachable
    exactly when it matters. Same reasoning as Phase 5's review route and every reversal route
    since.
    """
    position = cash_service.shift_cash_position(db, shift=access.shift)
    return CashPositionResponse(**vars(position))
